controllers/function_controller.go

```diff
@@ -91,26 +96,41 @@ func (r *FunctionReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		return ctrl.Result{}, err
 	}
 
-	err = r.ApplyFunctionStatefulSet(ctx, function)
+	isNewGeneration := r.checkIfFunctionGenerationsIsIncreased(function)
+
+	err = r.ApplyFunctionStatefulSet(ctx, function, isNewGeneration)
 	if err != nil {
 		return reconcile.Result{}, err
 	}
-	err = r.ApplyFunctionService(ctx, req, function)
+	err = r.ApplyFunctionService(ctx, function, isNewGeneration)
 	if err != nil {
 		return reconcile.Result{}, err
 	}
-	err = r.ApplyFunctionHPA(ctx, req, function)
+	err = r.ApplyFunctionHPA(ctx, function, isNewGeneration)
 	if err != nil {
 		return reconcile.Result{}, err
 	}
 
+	r.functionGenerations.Store(spec.GetNamespacedName(function, v1alpha1.FunctionComponent), function.Generation)
 	return ctrl.Result{}, nil
 }
 
+func (r *FunctionReconciler) checkIfFunctionGenerationsIsIncreased(function *v1alpha1.Function) bool {
```

Comment:
seems like the cached `functionGenerations` will always trigger as a new generation when the controller restart. Is it desired? or can we move the observed generation to the function's status? like add an `ObservedGeneration` to `FunctionStatus`?

Reply:
please refer to https://github.com/streamnative/pulsar-operators/pull/365 for more details about the idea about `ObservedGeneration`

Reply:
good idea, I will update it.

Reply:
done